packages/runtime/container-runtime/src/test/containerRuntime.spec.ts

```diff
@@ -445,6 +446,90 @@ describe("Runtime", () => {
 					}
 				});
+
+			for (const skipSafetyFlushDuringProcessStack of [true, undefined]) {
+				it("Inbound (non-runtime) op triggers flush due to refSeq changing", async () => {
+					const submittedBatches: {
+						messages: IBatchMessage[];
+						referenceSequenceNumber: number;
+					}[] = [];
+
+					const mockContext = getMockContext({
+						settings: {
+							"Fluid.ContainerRuntime.DisableFlushBeforeProcess":
+								skipSafetyFlushDuringProcessStack,
+						},
+					});
+					(
+						mockContext as { submitBatchFn: IContainerContext["submitBatchFn"] }
+					).submitBatchFn = (
+						messages: IBatchMessage[],
+						referenceSequenceNumber: number = -1,
+					) => {
+						submittedOps.push(...messages); // Reusing submittedOps since submitFn won't be invoked due to submitBatchFn's presence
+						submittedBatches.push({ messages, referenceSequenceNumber });
+						return 999; // CSN not used in test asserts below
+					};
+
+					const containerRuntime = await ContainerRuntime.loadRuntime({
+						context: mockContext as IContainerContext,
+						registryEntries: [],
+						existing: false,
+						runtimeOptions: {},
+						provideEntryPoint: mockProvideEntryPoint,
+					});
+
+					// Define resubmit and setConnectionState on channel collection
+					// defineResubmitAndSetConnectionState(containerRuntime);
+
+					// Submit the first message
+					submitDataStoreOp(containerRuntime, "1", "testMessage1");
+					assert.strictEqual(submittedOps.length, 0, "No ops submitted yet");
+
+					// Bump lastSequenceNumber and trigger the "op" event artificially to simulate processing a non-runtime op
+					const mockDeltaManager = mockContext.deltaManager as MockDeltaManager;
+					++mockDeltaManager.lastSequenceNumber;
+					mockDeltaManager.emit("op", {
+						clientId: mockClientId,
+						sequenceNumber: mockDeltaManager.lastSequenceNumber,
+						clientSequenceNumber: 1,
+						type: MessageType.Operation,
+						contents: JSON.stringify({
+							type: "nonRuntimeOp",
+							contents: "nonRuntimeOpContent",
+						}),
+					});
+
+					const expectedSubmitCount = skipSafetyFlushDuringProcessStack ? 0 : 1;
+					assert.strictEqual(
+						submittedOps.length,
+						expectedSubmitCount,
+						"Submitted op count wrong after first op",
+					);
+
+					// Submit some more messages.  If we didn't flush on "op", this would throw
+					// Then wait for flush to get triggered, will flush only the 2nd message
+					// Note: We have to submit multiple here to trigger maybeSubmitPartialBatch in the case skipSafetyFlushDuringProcessStack: true
+					submitDataStoreOp(containerRuntime, "2", "testMessage2");
+					submitDataStoreOp(containerRuntime, "3", "testMessage3");
+					await Promise.resolve();
+
+					// Validate that the messages were submitted (as a grouped batch)
+					assert.strictEqual(submittedOps.length, 2, "Two messages should be submitted");
+					assert(
```

Comment:
Is this assert really needed? It's getting into the internals of the content. It might be better to validate that the second op has 2 ops inside of it (if possible)

Reply:
The third op wasn't needed, removed this